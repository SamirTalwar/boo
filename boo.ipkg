package boo

sourcedir = "proofs"

modules =
    Builtins
  , EvaluationReduction
  , Expression
  , Identifier
  , Native
  , Primitive
  , Test
