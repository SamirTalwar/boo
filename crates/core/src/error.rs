use crate::span::Span;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, Clone, PartialEq, thiserror::Error, miette::Diagnostic)]
pub enum Error {
    #[error("Unexpected token: {token}")]
    #[diagnostic(code(boo::lex::unexpected_token))]
    UnexpectedToken {
        #[label("unexpected token")]
        span: Span,
        token: String,
    },

    #[error("Parse error: expected one of {expected_tokens:?}")]
    #[diagnostic(code(boo::parse::error))]
    ParseError {
        #[label("{}", expected_one_of(expected_tokens))]
        span: Span,
        expected_tokens: Vec<&'static str>,
    },

    #[error("Unknown variable: {name:?}")]
    #[diagnostic(code(boo::interpret::unknown_variable))]
    UnknownVariable {
        #[label("unknown variable")]
        span: Span,
        name: String,
    },

    #[error("Could not apply the function")]
    #[diagnostic(code(boo::interpret::unknown_variable))]
    InvalidFunctionApplication {
        #[label("invalid function")]
        span: Span,
    },
}

fn expected_one_of(strings: &[&str]) -> String {
    match strings {
        [] => "<nothing>".to_string(),
        [a] => format!("expected {}", a),
        [a, b] => format!("expected {} or {}", a, b),
        _ => format!(
            "expected one of {}, or {}",
            strings[0..strings.len() - 1].join(", "),
            strings[strings.len() - 1]
        ),
    }
}
