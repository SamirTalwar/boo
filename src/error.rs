#[derive(Debug, PartialEq)]
pub enum BooError {
    ParseError(peg::error::ParseError<peg::str::LineCol>),
}

impl std::fmt::Display for BooError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ParseError(error) => write!(f, "Parse error: {}", error),
        }
    }
}
