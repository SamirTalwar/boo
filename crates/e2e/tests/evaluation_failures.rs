use boo::error::{Error, Result};
use boo::*;
use boo_naive_evaluator::naively_evaluate;

#[test]
fn test_unknown_variable() -> Result<()> {
    expect_error(
        "unknown_variable",
        "123 + xyz",
        Error::UnknownVariable {
            span: (6..9).into(),
            name: "xyz".to_string(),
        },
    )
}

#[test]
fn test_does_not_close_over_variables_out_of_scope() -> Result<()> {
    expect_error(
        "does_not_close_over_variables_out_of_scope",
        "let fun = (let one = 1 in fn param -> one + param + external) in let external = 2 in fun 3",
        Error::UnknownVariable {
            span: (52..60).into(),
            name: "external".to_string(),
        },
    )
}

fn expect_error(name: &str, program: &str, expected_error: Error) -> Result<()> {
    let ast = parse(program)?;
    insta::with_settings!({ description => program }, {
        insta::assert_debug_snapshot!(name.to_string() + "__parse", ast);
    });

    let expr = boo::builtins::prepare(ast);
    let efficient_result = evaluate(expr.clone());
    assert_eq!(efficient_result, Err(expected_error.clone()));

    let naive_result = naively_evaluate(expr);
    assert_eq!(naive_result, Err(expected_error));

    Ok(())
}
