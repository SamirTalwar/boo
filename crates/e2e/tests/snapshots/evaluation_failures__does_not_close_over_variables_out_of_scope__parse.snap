---
source: crates/e2e/tests/evaluation_failures.rs
description: let fun = (let one = 1 in fn param -> one + param + external) in let external = 2 in fun 3
expression: ast
---
Expr(
    Spanned {
        span: Span {
            start: 0,
            end: 90,
        },
        value: Assign(
            Assign {
                name: Name(
                    "fun",
                ),
                value: Expr(
                    Spanned {
                        span: Span {
                            start: 11,
                            end: 60,
                        },
                        value: Assign(
                            Assign {
                                name: Name(
                                    "one",
                                ),
                                value: Expr(
                                    Spanned {
                                        span: Span {
                                            start: 21,
                                            end: 22,
                                        },
                                        value: Primitive(
                                            Integer(
                                                Small(
                                                    1,
                                                ),
                                            ),
                                        ),
                                    },
                                ),
                                inner: Expr(
                                    Spanned {
                                        span: Span {
                                            start: 26,
                                            end: 60,
                                        },
                                        value: Function(
                                            Function {
                                                parameter: Name(
                                                    "param",
                                                ),
                                                body: Expr(
                                                    Spanned {
                                                        span: Span {
                                                            start: 38,
                                                            end: 60,
                                                        },
                                                        value: Apply(
                                                            Apply {
                                                                function: Expr(
                                                                    Spanned {
                                                                        span: Span {
                                                                            start: 38,
                                                                            end: 60,
                                                                        },
                                                                        value: Apply(
                                                                            Apply {
                                                                                function: Expr(
                                                                                    Spanned {
                                                                                        span: Span {
                                                                                            start: 38,
                                                                                            end: 60,
                                                                                        },
                                                                                        value: Identifier(
                                                                                            Operator(
                                                                                                "+",
                                                                                            ),
                                                                                        ),
                                                                                    },
                                                                                ),
                                                                                argument: Expr(
                                                                                    Spanned {
                                                                                        span: Span {
                                                                                            start: 38,
                                                                                            end: 49,
                                                                                        },
                                                                                        value: Apply(
                                                                                            Apply {
                                                                                                function: Expr(
                                                                                                    Spanned {
                                                                                                        span: Span {
                                                                                                            start: 38,
                                                                                                            end: 49,
                                                                                                        },
                                                                                                        value: Apply(
                                                                                                            Apply {
                                                                                                                function: Expr(
                                                                                                                    Spanned {
                                                                                                                        span: Span {
                                                                                                                            start: 38,
                                                                                                                            end: 49,
                                                                                                                        },
                                                                                                                        value: Identifier(
                                                                                                                            Operator(
                                                                                                                                "+",
                                                                                                                            ),
                                                                                                                        ),
                                                                                                                    },
                                                                                                                ),
                                                                                                                argument: Expr(
                                                                                                                    Spanned {
                                                                                                                        span: Span {
                                                                                                                            start: 38,
                                                                                                                            end: 41,
                                                                                                                        },
                                                                                                                        value: Identifier(
                                                                                                                            Name(
                                                                                                                                "one",
                                                                                                                            ),
                                                                                                                        ),
                                                                                                                    },
                                                                                                                ),
                                                                                                            },
                                                                                                        ),
                                                                                                    },
                                                                                                ),
                                                                                                argument: Expr(
                                                                                                    Spanned {
                                                                                                        span: Span {
                                                                                                            start: 44,
                                                                                                            end: 49,
                                                                                                        },
                                                                                                        value: Identifier(
                                                                                                            Name(
                                                                                                                "param",
                                                                                                            ),
                                                                                                        ),
                                                                                                    },
                                                                                                ),
                                                                                            },
                                                                                        ),
                                                                                    },
                                                                                ),
                                                                            },
                                                                        ),
                                                                    },
                                                                ),
                                                                argument: Expr(
                                                                    Spanned {
                                                                        span: Span {
                                                                            start: 52,
                                                                            end: 60,
                                                                        },
                                                                        value: Identifier(
                                                                            Name(
                                                                                "external",
                                                                            ),
                                                                        ),
                                                                    },
                                                                ),
                                                            },
                                                        ),
                                                    },
                                                ),
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
                inner: Expr(
                    Spanned {
                        span: Span {
                            start: 65,
                            end: 90,
                        },
                        value: Assign(
                            Assign {
                                name: Name(
                                    "external",
                                ),
                                value: Expr(
                                    Spanned {
                                        span: Span {
                                            start: 80,
                                            end: 81,
                                        },
                                        value: Primitive(
                                            Integer(
                                                Small(
                                                    2,
                                                ),
                                            ),
                                        ),
                                    },
                                ),
                                inner: Expr(
                                    Spanned {
                                        span: Span {
                                            start: 85,
                                            end: 90,
                                        },
                                        value: Apply(
                                            Apply {
                                                function: Expr(
                                                    Spanned {
                                                        span: Span {
                                                            start: 85,
                                                            end: 88,
                                                        },
                                                        value: Identifier(
                                                            Name(
                                                                "fun",
                                                            ),
                                                        ),
                                                    },
                                                ),
                                                argument: Expr(
                                                    Spanned {
                                                        span: Span {
                                                            start: 89,
                                                            end: 90,
                                                        },
                                                        value: Primitive(
                                                            Integer(
                                                                Small(
                                                                    3,
                                                                ),
                                                            ),
                                                        ),
                                                    },
                                                ),
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
            },
        ),
    },
)
